### ISSUE TITLE
nvfortran fails to compile MPAS-A dycore due to generic operator interface issues #563

### ISSUE THREAD

**Issue by mgduda** [ESCOMP/CAM#563]
Compilation of two MPAS-A dycore source files fails with the `nvfortran` compiler due to apparent issues with the use of generic operator interfaces for derived types in MPAS's timekeeping module.

The error messages are as follows.

```
NVFORTRAN-S-0099-Illegal use of operator - on a derived type (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 187)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 187)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 187)
NVFORTRAN-S-0148-Reference to derived type required (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 187)
  0 inform,   0 warnings,   4 severes, 0 fatal for mpas_atm_update_bdy_tend
NVFORTRAN-S-0099-Illegal use of operator - on a derived type (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 376)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 376)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 376)
NVFORTRAN-S-0148-Reference to derived type required (components/cam/src/dynamics/mpas/dycore/src/core_atmosphere/dynamics/mpas_atm_boundaries.F: 376)
  0 inform,   0 warnings,   4 severes, 0 fatal for mpas_atm_get_bdy_state
```

```
NVFORTRAN-S-0099-Illegal use of operator + on a derived type (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2255)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2255)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2255)
NVFORTRAN-S-0148-Reference to derived type required (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2255)
NVFORTRAN-S-0099-Illegal use of operator < on a derived type (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2257)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2257)
NVFORTRAN-S-0099-Illegal use of derived type (components/cam/src/dynamics/mpas/driver/cam_mpas_subdriver.F90: 2257)
  0 inform,   0 warnings,   7 severes, 0 fatal for cam_mpas_run
```

The `-`, `+`, and `<` operators are defined in [`mpas_timekeeping.F`](https://github.com/MPAS-Dev/MPAS-Model/blob/develop/src/framework/mpas_timekeeping.F) as
```
   interface operator (+)
      module procedure add_t_ti
      module procedure add_ti_ti
   end interface

   interface operator (-)
      module procedure sub_t_t
      module procedure sub_t_ti
      module procedure sub_ti_ti
      module procedure neg_ti
   end interface

   interface operator (.LT.)
      module procedure lt_t_t
      module procedure lt_ti_ti
   end interface
```

**Possible solution:**
---------------------------
Explicitly importing the specific implementations of operators that are needed in the `mpas_atm_boundaries.F` and `cam_mpas_subdriver.F` files resolves the issue.

Since this issue only occurs with the `nvfortran` compiler (and, apparently only when MPAS-A is used as a dycore in CAM but not for stand-alone MPAS-A), enclosing Fortran `use` statements in preprocessing directives seems appropriate. For example, in `mpas_atm_boundaries.F`, adding
```
#ifdef __NVCOMPILER
       !
       ! Some versions of the nvfortran compiler complain about the illegal use
       ! of an operator on a derived type if the following specific
       ! implementation of the (-) operator is not explicitly imported
       !
    use mpas_timekeeping, only : sub_t_t
#endif
```
allows the file to be successfully compiled with `nvfortran` and has no impact on the code seen by other compilers.

**Comment by jtruesdal** [comment]
@mgduda if you can give me a new hash with this fix I will update the Externals and include this as part of my misc PR.

**Comment by mgduda** [comment]
@jtruesdal I think what we'll ultimately need are the following:
1. A new hash from the `MPAS-Model` repository from the 'develop' branch after including a fix in `mpas_atm_boundaries.F`
2. A fix in the `cam_mpas_subdriver.F` file, which resides in the CAM repository
3. A change to the MPAS-A dyore `Makefile` to accommodate recent changes to the MPAS-Model `develop` branch

Worth pointing out here is that we'll be updating the MPAS-A dycore more generally by including a hash from the HEAD of the MPAS-Model `develop` branch.

I'll let you know once we've gotten the fix to `mpas_atm_boundaries.F` merged to the MPAS-Model `develop` branch; this could take a day or two.

**Comment by mgduda** [comment]
@jtruesdal We've merged MPAS-Dev/MPAS-Model#973, and I now have a [commit](https://github.com/mgduda/CAM/commit/06346b8c) on the [mpas_nvhpc_operator_import](https://github.com/mgduda/CAM/commits/mpas_nvhpc_operator_import) branch in my fork of the CAM repository with the changes that I think we'll need in CAM.

**Comment by jtruesdal** [comment]
Thanks.  I’ll merge this into my misc PR and start testing
